+nosnoc/+discrete_problem/stewart.m

```diff
@@ -0,0 +1,771 @@
+                        nu = pi_lambda + pi_theta;
+                        eta = 1;
+                        for jjj=1:length(nu)
+                            eta = eta*nu(jjj);
+                        end
+                        eta_vec = [eta_vec;eta];
+                        delta_h = obj.w.h(ii,jj) - obj.w.h(ii,jj-1);
+                        obj.f = obj.f + obj.p.rho_h_p() * eta * delta_h.^2
+                    end
+                end
+              case 'direct'
+                eta_vec = [];
+                for ii=1:opts.N_stages
+                    for jj=2:opts.N_finite_elements(1)
+                        sigma_lambda_B = sum2(obj.w.lambda(ii,jj-1,:));
+                        sigma_theta_B = sum2(obj.w.theta(ii,jj-1,:));
+                        
+                        sigma_lambda_F = sum2(obj.w.lambda(ii,jj,:));
+                        sigma_theta_F = sum2(obj.w.theta(ii,jj,:));
+
+                        pi_lambda = sigma_lambda_B .* sigma_lambda_F;
+                        pi_theta = sigma_theta_B .* sigma_theta_F;
+                        nu = pi_lambda + pi_theta;
+                        eta = 1;
+                        for jjj=1:length(nu)
+                            eta = eta*nu(jjj);
+                        end
+                        eta_vec = [eta_vec;eta];
+                        delta_h = obj.w.h(ii,jj) - obj.w.h(ii,jj-1);
+                        obj.g.step_equilibration(ii,jj) = {eta*delta_h, 0, 0};
+                    end
+                end
+                obj.eta_fun = Function('eta_fun', {obj.w.w}, {eta_vec});
+              case 'direct_homotopy'
+                error("not currently implemented")
+                eta_vec = [];
+                for ii=1:opts.N_stages
+                    for jj=2:opts.N_finite_elements(1)
+                        sigma_lambda_B = sum2(obj.w.lambda(ii,jj-1,:));
+                        sigma_theta_B = sum2(obj.w.theta(ii,jj-1,:));
+                        
+                        sigma_lambda_F = sum2(obj.w.lambda(ii,jj,:));
+                        sigma_theta_F = sum2(obj.w.theta(ii,jj,:));
+
+                        pi_lambda = sigma_lambda_B .* sigma_lambda_F;
+                        pi_theta = sigma_theta_B .* sigma_theta_F;
+                        nu = pi_lambda + pi_theta;
+                        eta = 1;
+                        for jjj=1:length(nu)
+                            eta = eta*nu(jjj);
+                        end
+                        eta_vec = [eta_vec;eta];
+                        obj.eta_vec = eta_vec;
+                        delta_h = obj.w.h(ii,jj) - obj.w.h(ii,jj-1);
+                        homotopy_eq = [eta*delta_h - sigma;eta*delta_h + sigma];
+                        obj.g.step_equilibration(ii,jj) = {homotopy_eq, [-inf;0], [0;inf]};
+                    end
+                end
+                obj.eta_fun = Function('eta_fun', {obj.w.w}, {eta_vec});
+              case 'mlcp'
```

Comment:
mlcp is what? Please use a more instructive name.

Reply:
i guess it is the linear case? 
at some point if it turns out that the linear is always better than the old nonlinear ones, we should just delete the old nonlinear (direct, direct homotopy...)

Reply:
Would `linear` be a better name? I think in some sense `lcp`/`mlcp` is less deceptive as it explicitly points out that it formulates step-eq as a complementarity problem.